source/validate_id.cpp

```diff
@@ -710,11 +710,196 @@ bool idUsage::isValid<SpvOpSpecConstantFalse>(const spv_instruction_t* inst,
   return true;
 }
 
-#if 0
 template <>
-bool idUsage::isValid<SpvOpSpecConstantComposite>(
-    const spv_instruction_t *inst, const spv_opcode_desc opcodeEntry) {}
-#endif
+bool idUsage::isValid<SpvOpSpecConstantComposite>(const spv_instruction_t* inst,
+                                                  const spv_opcode_desc) {
+  // The result type must be a composite type.
+  auto resultTypeIndex = 1;
+  auto resultType = module_.FindDef(inst->words[resultTypeIndex]);
+  if (!resultType || !spvOpcodeIsComposite(resultType->opcode())) {
+    DIAG(resultTypeIndex) << "OpSpecConstantComposite Result Type <id> '"
+                          << inst->words[resultTypeIndex]
+                          << "' is not a composite type.";
+    return false;
+  }
+  // Validation checks differ based on the type of composite type.
+  auto constituentCount = inst->words.size() - 3;
+  switch (resultType->opcode()) {
+    // For Vectors, the following must be met:
+    // * Number of constituents in the result type and the vector must match.
+    // * All the components of the vector must have the same type (or specialize
```

Comment:
This comment is perhaps not well-written; I apologize. I should rewrite the comment.
What this code does is that it ensures that each of the operands of OpSpecConstantComposite has the proper type. For example, for the code below, it ensures that %f1 and %f2 and %f3 and %f4 have the type that %vec4 components should have (i.e. ensures they're all float). There is no instruction before OpSpecConstantComposite that could validate this.

```
%floatt = OpTypeFloat 32
%vec4  = OpTypeVector %floatt 4
%f1       = OpSpecConstant %floatt 1.0
%f2       = OpSpecConstant %floatt 2.0
%f3       = OpSpecConstant %floatt 3.0
%f4       = OpSpecConstant %floatt 4.0
%spec_constt = OpSpecConstantComposite %vec4 %f1 %f2 %f3 %f4
```

Does that make sense? Or am I missing your point?